Adds migration for new settings and settings integrity test

rshbhgrg opened the pull request:
Added new settings integrity tests

- We made changes to settings table structure in 3.22 which added new columns for `group` and `flags`
- Any new setting needs explicit group and flag migrations since or they fallback to default group value of `core`
- PR adds 2 new tests to ensure
  - settings hash need to be updated in DB integrity test anytime default-settings are changed
  - a migration is present by maintaining a whitelist of allowed core settings and failing if new setting is added without correct group migration

Added migration to update group for new portal settings 

- 3 new portal settings were added in `3.23` in default settings - `portal_button_style`, `portal_button_icon`, `portal_button_signup_text`
- New settings default to group `core` on migrating from pre 3.22 versions due to missing columns in DB
- Migration here updates the new settings to correct group

rishabhgrg commented on test/unit/data/schema/integrity_spec.js:
```diff
@@ -42,3 +45,62 @@ describe('DB version integrity', function () {
+describe('Settings integrity', function () {
+    let ghostServer;
+    before(function () {
+        return ghost()
```
Unsure if its the right approach to do ghost start in unit tests as we don't have precedence for it, but need to do this here to initialize and test settings in DB, interested in any counter input/thoughts on this or if I missed a better approach. 🤔

allouis replied:
Yeah, this isn't a unit test if it has to start the whole of Ghost :confused: 

I would either move this test out of the unit tests, OR use a more "unit-ey"  approach of hashing the file and checking it - like we do above for the fixtures & schema

rishabhgrg replied:
Thanks @allouis, have re-written integrity tests as 2 separate checks, one in unit which checks for file hash changes, and other which checks correct values are added to DB.

allouis commented on core/server/data/migrations/versions/3.24/01-populate-group-for-new-portal-settings.js:
```diff
@@ -0,0 +1,39 @@
+const logging = require('../../../../../shared/logging');
+
+// new setting keys and group mapping
+const groupMapping = [{
```
I don't think there's a need to have this as a one item array? NABD tho

rishabhgrg replied:
Same as [here](https://github.com/TryGhost/Ghost/pull/12034#discussion_r452760978), the primary reason for this structure of test is allowing easier reuse in future for n number of new settings with different groups, that only need to update this mapping.

allouis replied:
cool okay

allouis commented on core/server/data/migrations/versions/3.24/01-populate-group-for-new-portal-settings.js:
```diff
@@ -0,0 +1,39 @@
+            return await Promise.map(groupMap.keys, async (key) => {
+                logging.info(`Updating setting ${key} to group ${groupMap.group}`);
+
+                return await options
```
I feel like this whole `up` method could be simplified to
```js
await options
  .transacting('settings')
  .update({
    group: 'portal'
  })
  .whereIn('key', ['portal_button_style', 'portal_button_icon', 'portal_button_signup_text');
```

rishabhgrg replied:
@allouis Primary reason for structuring it like this is allowing future settings migration for group/flags to reuse this easily and just update the mapping on top

naz replied:
@rishabhgrg I'm siding with @allouis on this one. You could still keep the mapping on top and then use something like:
```js
await options
  .transacting('settings')
  .update({
    group: groupMap.gorup
  })
  .whereIn('key', groupMap.keys);
```
At least that would limit the nesting to one level and keep the configuration more-less generic :thinking:

allouis commented on core/server/data/migrations/versions/3.24/01-populate-group-for-new-portal-settings.js:
```diff
@@ -0,0 +1,39 @@
+    // `up` is only run to update correct group value for each setting instead of default.
+    // it doesn't make sense to be revert a setting's group to default `core` again
+    async down() {
+        return Promise.resolve();
```
NABD but `async` functions already return a promise, so you can leave this totally blank

rshbhgrg committed:
Added migration to update group for new portal settings

closes https://github.com/TryGhost/Ghost/issues/12026

- 3 new portal settings were added in `3.23` in default settings - `portal_button_style`, `portal_button_icon`, `portal_button_signup_text`
- New settings default to group `core` on migrating from pre 3.22 versions due to missing columns in DB
- Migration here updates the new settings to correct group

rshbhgrg committed:
Added new settings integrity tests

refs https://github.com/TryGhost/Ghost/issues/12026

- We made changes to settings table structure in 3.22 which added new columns for `group` and `flags`
- Any new setting needs explicit group and flag migrations since or they fallback to default group value of `core`
- This tests ensures
  - hash is updated in DB integrity test anytime default-settings are changed
  - that a migration is present by maintaining a whitelist of allowed core settings and failing if new setting is added without correct group migration

allouis commented on test/regression/settings/settings_spec.js:
```diff
@@ -0,0 +1,55 @@
+const db = require('../../../core/server/data/db');
+const testUtils = require('../../utils');
+const ghost = testUtils.startGhost;
```
`ghost` is never used anywhere

allouis commented on test/regression/settings/settings_spec.js:
```diff
@@ -0,0 +1,55 @@
+ */
+
+describe('Settings', function () {
+    let ghostServer;
```
`ghostServer` is never used anywhere

allouis commented on test/regression/settings/settings_spec.js:
```diff
@@ -0,0 +1,55 @@
+                throw err;
+            });
+    });
+});
```
Missing newline :stuck_out_tongue:

allouis reviewed (commented):
Looks good! I think test is much nicer split out!

daniellockyer commented on test/unit/data/schema/integrity_spec.js:
```diff
@@ -3,31 +3,36 @@ const _ = require('lodash');
  *
  * 1. A new blog get's installed and the database looks correct and complete.
  * 2. A blog get's updated from a lower Ghost version and the database looks correct and complete.
+ * 3.
```
but what is 3? :grin:

rishabhgrg replied:
Haha, realized half-way its not needed :P Updated
